### PR TITLE
Infini attention blog post

### PR DESCRIPTION
**Pull request by xrsrke**

### CODE REVIEW

File: infini-attention.md [hunk 1]
```diff
@@ -0,0 +1,295 @@
+---
+title: "A failed experiment: Infini-Attention, and why we should keep trying?" 
+thumbnail: ./assets/185_infini_attention/infini_attention_thumbnail.png
+authors:
+- user: xrsrke
```
Comment (pcuenca) [inline infini-attention.md:5]:
```suggestion
- user: zzhhjjj
```

This should be a HF user handle :)

**Review by pcuenca — COMMENTED** [review]
Hi @xrsrke! Thanks a lot for working on this, do let us know when you want a review :) Meanwhile, a couple of comments about the structure, just in case you didn't know:
- You need to add a metadata entry to `_blog.yml` before publishing.
- All images, except the thumbnail, are usually stored in [this dataset](https://huggingface.co/datasets/huggingface/documentation-images/tree/main/blog) to keep this repo as lean as possible.

**Comment by xrsrke** [comment]
Hi @pcuenca thanks for the info. I've removed all the other images from the repo, and move to here https://huggingface.co/datasets/huggingface/documentation-images/discussions/352. It's ready for review!

**Review by osanseviero — COMMENTED** [review]
Looking very nice! :fire:

File: infini-attention.md [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+- user: lvwerra
+- user: thomwolf
+---
+
+The context length of language models is one of the central attributes besides the model’s performance. Especially since the emergence of in-context learning it has become more and more important to add relevant information to the model’s input. Thus the context length rapidly increased from paragraph (512 tokens of BERT/GPT-1) to pages (1024/2048 of GPT-2 and GPT-3 respectively) to books (128k of Claude) all the way to collections of books (1-10M tokens of Gemini). However extending standard attention to such length remains challenging.
```
[the remark this answers is not in the document]
Comment (xrsrke) [reply infini-attention.md]: added to the "Recap on the memory complexity of self-attention" section

File: infini-attention.md [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+Motivated by this, we explore an alternative approach to standard attention: infini-attention. The paper was released by researchers from Google in April 2024 [[link]](https://arxiv.org/abs/2404.07143). Instead of computing attention scores between every word, Infini attention divides the sequence into segments, compresses earlier segments into a fixed buffer, and allows the next segment to retrieve memory from the earlier segments while limiting attention scores to words within the current segment.  A key advantage is that it uses the same query within a segment to access information from both its own segment and the compressed memory. This enables us to cheaply extend the context length for a pretrained model. Since it updates the compressed memory on the go, and success is not guaranteed, but if it all works out we can have infinite context length, as it only keeps a single buffer for all the memory of earlier segments. However, by definition, compression has limits on the amount of information it can effectively compress. The question now is: how good is the compressed memory if it works?
```
[the remark this answers is not in the document]
Comment (xrsrke) [reply infini-attention.md]: `[[link]](url)` seems to be better? because it will render something like this [[link]](https://google.com)

[the remark this answers is not in the document]
Comment (xrsrke) [reply infini-attention.md]: good idea. thanks. will use "Infini-attention" as the same as in the paper

File: infini-attention.md [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+
+However, conceptually understanding a new method can be relatively easy compared to actually making it work, and this is rarely shared publicly (we spent 90% of our time debugging the convergence issue). Motivated by this, we want to share how we reproduced the Infini-attention paper, what motivated us throughout the debugging process, and how hard it is to make these things work.
+
+Following the recent release of Llama 3 8B, which has a context length limit of 8k tokens, we sought to extend this length to 1m tokens without quadratically increasing the memory. In this blog post, we will start by explaining how Infini Attention works. We’ll then outline our reproduction principles and describe our initial small-scale experiment. We will discuss the challenges we faced, how we addressed them, and conclude with a summary of our findings and other ideas we explored. If you’re interested in testing our trained checkpoint, you can find it in the following repo [[link]](https://github.com/huggingface/nanotron/tree/xrsrke/this_infini_version_work) (note that, as the technique doesn’t work well enough, so we did not invest much effort in cleaning up the code).
```
[the remark this answers is not in the document]
Comment (xrsrke) [reply infini-attention.md]: since our experiments was using llama3 8b, will remove "recent release"

**Comment by xrsrke** [comment]
hi @osanseviero. thanks for the suggestions, all fixed, could u please review it? thanks

File: _blog.yml [hunk 5]
```diff
@@ -4528,3 +4527,12 @@
     - LLM
     - nlp
     - community
+- local: infini-attention
+  title: "A failed experiment: Infini-Attention, and why we should keep trying?"
+  author: neuralink
+  thumbnail: /blog/assets/quanto-diffusers/thumbnail.png
+  date: August 5, 2024
```
Comment (osanseviero) [inline _blog.yml:4534]: to update

File: infini-attention.md [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+title: "A failed experiment: Infini-Attention, and why we should keep trying?" 
+thumbnail: ./assets/185_infini_attention/infini_attention_thumbnail.png
+authors:
+- user: neuralink
+- user: lvwerra
+- user: thomwolf
+---
+
+
+TLDR: Infini-attention's performance gets worse as we increase the number of times we compress the memory, and to the best of our knowledge, ring attention [[link]](https://x.com/Haojun_Zhao14/status/1815419356408336738), YaRN [[link]](https://arxiv.org/abs/2309.00071) and rope scaling [[link]](https://arxiv.org/abs/2309.16039) are still the best ways for extending a pretrained model to longer context length.
```
Comment (osanseviero) [inline infini-attention.md:11]:
```suggestion
TLDR: Infini-attention's performance gets worse as we increase the number of times we compress the memory, and to the best of our knowledge, [ring attention](https://x.com/Haojun_Zhao14/status/1815419356408336738), [YaRN](https://arxiv.org/abs/2309.00071) and [rope scaling](https://arxiv.org/abs/2309.16039) are still the best ways for extending a pretrained model to longer context length.
```

File: infini-attention.md [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+
+
+# Section 0: Introduction
+
+The context length of language models is one of the central attributes besides the model’s performance. Since the emergence of in-context learning, adding relevant information to the model’s input has become increasingly important. Thus, the context length rapidly increased from paragraphs (512 tokens with BERT/GPT-1) to pages (1024/2048 with GPT-2 and GPT-3 respectively) to books (128k of Claude) all the way to collections of books (1-10M tokens of Gemini). However, extending standard attention to such length remains challenging.
+
+> A little into to Ring Attention: to the best of our knowledge, Ring Attention was first introduced by researchers from UC Berkeley in 2024 [[link]](https://arxiv.org/abs/2310.01889). It is an engineering technique that helps overcome memory limitations by allowing us to obtain an output equivalent to standard attention without storing all keys and values in a sequence.
```
Comment (osanseviero) [inline infini-attention.md:18]:
```suggestion
> A small intro to Ring Attention: to the best of our knowledge, Ring Attention was first introduced by researchers from UC Berkeley in 2024 [[link]](https://arxiv.org/abs/2310.01889). It is an engineering technique that helps overcome memory limitations by allowing us to obtain an output equivalent to standard attention without storing all keys and values in a sequence.
```

Comment (osanseviero) [inline infini-attention.md:18]: Let's consistently not capitalize ring attention for consistency

File: infini-attention.md [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+
+Even with ring attention, the number of GPUs required to train a Llama 3 8B [[link]](https://arxiv.org/abs/2407.21783) on a 1-million-token context length with a batch size of 1 is 512 GPUs [[link]](https://x.com/Haojun_Zhao14/status/1815419356408336738). As scaling law has shown [[link]](https://arxiv.org/abs/2001.08361), there is a strong correlation between model size and its downstream performance, which means the bigger the model, the better (of course, both models should be well-trained). So we not only want a 1m context length, but we want a 1m context length on the biggest model (e.g., Llama 3 8B 405B). And there are only a few companies in existence that have the resources to do so.
```
Comment (osanseviero) [inline infini-attention.md:20]: Please update links so they are in the text rather than as additional E.g.

```suggestion
Even with ring attention, the number of GPUs required to train a [Llama 3 8B](https://arxiv.org/abs/2407.21783) on a 1-million-token context length with a batch size of 1 is 512 GPUs. As scaling laws have [shown](https://arxiv.org/abs/2001.08361), there is a strong correlation between model size and its downstream performance, which means the bigger the model, the better (of course, both models should be well-trained). So we not only want a 1m context length, but we want a 1m context length on the biggest model (e.g., Llama 3 8B 405B). And there are only a few companies in existence that have the resources to do so.
```

File: infini-attention.md [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+    )
+
+    # NOTE: because normalization is the sum of all the keys, so each word should have the same normalization
+    retrieved_memory = retrieved_memory / denominator
+    return retrieved_memory
+```
+
+- Step 4: Combine the local context (from the current segment) with the long-term context (retrieved from the compressive memory) to generate the final output. This way, both short-term and long-term contexts can be considered in the attention output.
+
+    $$A=\operatorname{sigmoid}(\beta) \odot A_{\text {mem }}+(1-\operatorname{sigmoid}(\beta)) \odot A_{\text {dot }}$$
```
Comment (osanseviero) [inline infini-attention.md:88]: Are you sure operatorname is supported?

File: infini-attention.md [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+
+
+# Section 4: Convergable?
+
+So we simulated how the amount of weight would change during training given gradients are in a good range (L2 norm is 0.01), and found that, given the config of the last 8B LLaMA3 fine-tuning experiment, the total of absolute changes in the weight would be 0.03. Since we initialize balance factors at 0 (it doesn’t matter in this case), the weights at the end would be in the range [0 - 0.03, 0 + 0.03] = [-0.03, 0.03].
+
+An educated guess for ininfi attention to work well is its global weights spread out in the range 0 and 1 as in the paper. Given the weight above, sigmoid([-0.03, 0.03]) = tensor([0.4992, 0.5008]) (this fits with our previous experiment results that the balance factor is ~0.5). The next step is to use a higher learning rate for balance factors (and all other parameters use Llama 3 8B's learning rate), and a total number of training steps that allow the balance factors to change by at least 4, so that we allow global weights to reach the ideal weights if gradient descent wants (sigmoid(-4) ≈ 0, sigmoid(4) ≈ 1).
+
+![Figure 4a: generation](./assets/185_infini_attention/total_abs_change_in_adam.png)
```
Comment (osanseviero) [inline infini-attention.md:230]: The images need to be added to the Hub and updated here

File: infini-attention.md [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+
+**The results**
+
+![](./assets/185_infini_attention/200m_generation_first_signals.png)
+
+As you can see it somewhat works, and if you look at these sample generations, you can see that Infini-attention generates content related to the earlier segment.
+
+Since Infini-attention predicts the first token in the second segment by conditioning on the entire content of the first segment, which it generated as "_grad" for the first token, this provides a good signal. To validate whether this signal is a false positive, we hypothesize that Infini-attention generates content related to its earlier segment because when given "_grad" as the first generated token of the second segment, it consistently generates PyTorch-related tutorials, which happen to relate to its earlier segment. Therefore, we conducted a sanity test where the only input token was "_grad", and it generated [text here]. This suggests it does use the memory, but just doesn’t use it well enough (to retrieve the exact needle or continue the exact content of its earlier segment). The generation:
+
+> _graduate_education.html
```
Comment (osanseviero) [inline infini-attention.md:213]: To format

File: infini-attention.md (hunk 10 above)
Comment (xrsrke) [reply infini-attention.md:230]: I've already added it to the repo

image

**Comment by xrsrke** [comment]
Hi @osanseviero I had added all the requested change

**Review by osanseviero — APPROVED** [review]
